File: src/deploy/ocsort/python/ocsort_tracker.py
# Copyright 2026 SpacemiT (Hangzhou) Technology Co. Ltd.
#
# SPDX-License-Identifier: Apache-2.0

"""
OCSort Tracker (BaseModel style)

This wraps the existing OC-SORT python implementation and a YOLOv8 detector
into a factory-loadable model, similar to ByteTrack.
"""

from __future__ import annotations

from typing import Any, Dict, List, Optional, Tuple

import cv2
import numpy as np

from core import BaseModel
from deploy.yolov8 import YOLOv8Detector
from core.python.vision_model_exceptions import (
    ModelLoadError,
    ModelInferenceError,
)
from .tracking.ocsort import OCSort


class OCSortTracker(BaseModel):
    """
    OC-SORT tracker integrated with YOLOv8 detector.

    `track(image)` returns a list of dicts:
      - tlbr: [x1,y1,x2,y2] - bounding box coordinates
      - track_id: int - unique tracking ID
      - score: float - detection confidence score
      - class_id: int - object class ID
    """

    def __init__(
        self,
        model_path: str,
        # detector params
        conf_threshold: float = 0.25,
        iou_threshold: float = 0.45,
        num_threads: int = 4,
        # tracker params
        track_thresh: float = 0.6,  # maps to det_thresh
        track_iou_thresh: float = 0.3,
        track_buffer: int = 60,  # maps to max_age
        min_hits: int = 3,
        delta_t: int = 3,
        asso_func: str = "iou",
        inertia: float = 0.2,
        use_byte: bool = False,
        **kwargs,
    ):
        # Store parameters for detector initialization
        self.conf_threshold = conf_threshold
        self.iou_threshold = iou_threshold
        self.num_threads = num_threads

        # Store tracker parameters
        self.track_thresh = float(track_thresh)
        self.track_iou_thresh = float(track_iou_thresh)
        self.track_buffer = int(track_buffer)
        self.min_hits = int(min_hits)
        self.delta_t = int(delta_t)
        self.asso_func = str(asso_func)
        self.inertia = float(inertia)
        self.use_byte = bool(use_byte)

        # Detector will be initialized in _load_model
        self.detector: Optional[YOLOv8Detector] = None

        # OC-SORT tracker instance (pure python, no model loading needed)
        self.tracker = OCSort(
            det_thresh=self.track_thresh,
            max_age=self.track_buffer,
            min_hits=self.min_hits,
            iou_threshold=self.track_iou_thresh,
            delta_t=self.delta_t,
            asso_func=self.asso_func,
            inertia=self.inertia,
            use_byte=self.use_byte,
        )

        # Initialize BaseModel (will call _load_model if not lazy_load)
        super().__init__(model_path=model_path, lazy_load=kwargs.get('lazy_load', False), **kwargs)

    def _load_model(self, **kwargs):
        """
        Load YOLOv8 detection model.

        This method initializes the YOLOv8 detector which will be used for detection.
        """
        try:
            # Initialize YOLOv8 detector
            self.detector = YOLOv8Detector(
                model_path=self.model_path,
                conf_threshold=self.conf_threshold,
                iou_threshold=self.iou_threshold,
                num_threads=self.num_threads,
                **kwargs
            )

            # Save input shape from detector
            self.input_shape = self.detector.input_shape

        except Exception as e:
            raise ModelLoadError(
                model_path=self.model_path,
                reason=str(e),
                message=f"OCSort detector (YOLOv8) 加载失败: {e}"
            ) from e

    def preprocess(self, image: np.ndarray) -> Tuple[np.ndarray, float, Tuple[float, float]]:
        """
        Preprocess image using YOLOv8 detector's preprocessing.

        Args:
            image: Input image in BGR format

        Returns:
            Tuple of (preprocessed_image, ratio, pad) as returned by YOLOv8Detector.preprocess
        """
        self._ensure_model_loaded()
        return self.detector.preprocess(image)

    def postprocess(self, outputs: Any, **kwargs) -> Any:
        """
        Postprocess is not used directly in tracking.
        Detection postprocessing is handled by YOLOv8Detector.
        """
        # This method is required by BaseModel but not used in tracking flow
        # Detection postprocessing is done by detector.infer()
        pass

    def infer(self, image: np.ndarray) -> List[Dict[str, Any]]:
        """
        Run detection + OC-SORT tracking on a single frame.

        Args:
            image: Input image in BGR format

        Returns:
            List of tracking results, each containing:
            - tlbr: [x1, y1, x2, y2] bounding box coordinates
            - track_id: Unique tracking ID
            - score: Detection confidence score
            - class_id: Object class ID
        """
        self._ensure_model_loaded()

        # Run YOLOv8 detection
        try:
            detections = self.detector.infer(image)
        except Exception as e:
            raise ModelInferenceError(
                model_name="OCSortTracker",
                reason=f"检测失败: {e}",
                message=f"OCSort 检测错误: {e}"
            ) from e

        results: List[Dict[str, Any]] = []

        if detections is None or len(detections) == 0:
            # 没有检测结果，使用空数组更新追踪器以保持状态
            self.tracker.update_public(
                np.empty((0, 4), dtype=np.float32),
                np.empty((0,), dtype=np.int32),
                np.empty((0,), dtype=np.float32)
            )
            return results

        # Extract boxes, classes, scores from detections
        boxes = np.array([det["bbox"] for det in detections], dtype=np.float32)
        classes = np.array([det["class_id"] for det in detections], dtype=np.int32)
        scores = np.array([det["confidence"] for det in detections], dtype=np.float32)

        # Update OC-SORT tracker
        try:
            online = self.tracker.update_public(boxes, classes, scores)
        except Exception as e:
            raise ModelInferenceError(
                model_name="OCSortTracker",
                reason=f"追踪失败: {e}",
                message=f"OCSort 追踪错误: {e}"
            ) from e

        if online is None or len(online) == 0:
            return results

        # online: [x1,y1,x2,y2,id,cate,pad]
        for t in online:
            x1, y1, x2, y2 = map(float, t[:4])
            track_id = int(t[4])
            class_id = int(t[5]) if len(t) > 5 else 0

            # Find corresponding score from original detections
            score = 1.0
            for det in detections:
                if det["class_id"] == class_id:
                    # Try to match by position (simple heuristic)
                    det_bbox = det["bbox"]
                    if abs(det_bbox[0] - x1) < 5 and abs(det_bbox[1] - y1) < 5:
                        score = det["confidence"]
                        break

            results.append({
                "tlbr": [x1, y1, x2, y2],
                "track_id": track_id,
                "score": score,
                "class_id": class_id,
            })

        return results

    def track(self, image: np.ndarray) -> List[Dict[str, Any]]:
        """
        Alias for infer() method for consistency with ByteTrackTracker.

        Args:
            image: Input image in BGR format

        Returns:
            List of tracking results (same format as infer())
        """
        return self.infer(image)

    def draw_results(
        self,
        image: np.ndarray,
        results: List[Dict[str, Any]],
        labels: Optional[List[str]] = None,
    ) -> np.ndarray:
        """Simple visualization similar to ByteTrackTracker.draw_results()."""
        out = image.copy()
        if not results:
            return out
        colors = self._generate_colors(100)
        for r in results:
            tlbr = r.get("tlbr")
            if not tlbr or len(tlbr) != 4:
                continue
            x1, y1, x2, y2 = map(int, tlbr)
            track_id = int(r.get("track_id", -1))
            class_id = int(r.get("class_id", 0))
            score = float(r.get("score", 0.0))
            color = colors[track_id % len(colors)] if track_id >= 0 else (0, 255, 0)
            cv2.rectangle(out, (x1, y1), (x2, y2), color, 2)
            if labels and 0 <= class_id < len(labels):
                text = f"{labels[class_id]} ID:{track_id} {score:.2f}"
            else:
                text = f"ID:{track_id} {score:.2f}"
            (tw, th), bl = cv2.getTextSize(text, cv2.FONT_HERSHEY_SIMPLEX, 0.6, 2)
            cv2.rectangle(out, (x1, max(0, y1 - th - bl - 6)), (x1 + tw + 6, y1), color, -1)
            cv2.putText(out, text, (x1 + 3, y1 - 6), cv2.FONT_HERSHEY_SIMPLEX, 0.6, (255, 255, 255), 2)
        return out

    @staticmethod
    def _generate_colors(num_colors: int) -> List[tuple]:
        np.random.seed(42)
        return [tuple(map(int, np.random.randint(0, 255, 3))) for _ in range(num_colors)]

File: src/deploy/ocsort/cpp/ocsort_tracker.cpp
/*
 * Copyright (C) 2026 SpacemiT (Hangzhou) Technology Co. Ltd.
 * SPDX-License-Identifier: Apache-2.0
 */

#include "ocsort_tracker.h"

#include <cassert>
#include <chrono>
#include <algorithm>
#include <memory>
#include <stdexcept>
#include <string>
#include <variant>
#include <vector>

#include "common.h"
#include "vision_model_config.h"
#include "vision_model_factory.h"

namespace vision_deploy {

std::unique_ptr<vision_core::BaseModel> OCSortTracker::create(const YAML::Node& config, bool lazy_load) {
    std::string model_path = vision_core::yaml_utils::getString(config, "model_path");
    if (model_path.empty()) {
        throw std::runtime_error("model_path not found in config for OCSortTracker");
    }

    YAML::Node default_params = config["default_params"];
    if (!default_params) {
        throw std::runtime_error("default_params not found in config for OCSortTracker");
    }

    float conf_threshold = vision_core::yaml_utils::getFloat(default_params, "conf_threshold", 0.25f);
    float iou_threshold = vision_core::yaml_utils::getFloat(default_params, "iou_threshold", 0.45f);
    float det_thresh = vision_core::yaml_utils::getFloat(default_params, "det_thresh", 0.3f);
    int max_age = vision_core::yaml_utils::getInt(default_params, "max_age", 30);
    int min_hits = vision_core::yaml_utils::getInt(default_params, "min_hits", 3);
    int delta_t = vision_core::yaml_utils::getInt(default_params, "delta_t", 3);
    float inertia = vision_core::yaml_utils::getFloat(default_params, "inertia", 0.2f);
    bool use_byte = vision_core::yaml_utils::getBool(default_params, "use_byte", false);
    int num_threads = vision_core::yaml_utils::getInt(default_params, "num_threads", 4);
    std::string provider = vision_core::yaml_utils::getProvider(config);

    return std::make_unique<OCSortTracker>(
        model_path,
        conf_threshold,
        iou_threshold,
        det_thresh,
        max_age,
        min_hits,
        delta_t,
        inertia,
        use_byte,
        num_threads,
        lazy_load,
        provider);
}

OCSortTracker::OCSortTracker(const std::string& model_path,
                            float conf_threshold,
                            float iou_threshold,
                            float det_thresh,
                            int max_age,
                            int min_hits,
                            int delta_t,
                            float inertia,
                            bool use_byte,
                            int num_threads,
                            bool lazy_load,
                            const std::string& provider)
    : BaseModel(model_path, lazy_load),
        conf_threshold_(conf_threshold),
        iou_threshold_(iou_threshold),
        det_thresh_(det_thresh),
        max_age_(max_age),
        min_hits_(min_hits),
        delta_t_(delta_t),
        inertia_(inertia),
        use_byte_(use_byte),
        num_threads_(num_threads),
        provider_(provider) {
    // Initialize detector
    detector_ = std::make_unique<YOLOv8Detector>(
        model_path, conf_threshold_, iou_threshold_, num_threads_, lazy_load, provider_);

    // Initialize tracker (will be fully initialized after detector loads)
    if (!lazy_load) {
        load_model();
    }
}

void OCSortTracker::load_model() {
    if (model_loaded_) {
        return;
    }
    // Load detection model (idempotent if detector was eager-loaded in ctor).
    detector_->load_model();

    // Initialize OC-SORT tracker after detector is loaded
    tracker_ = std::make_unique<ocsort::OCSort>(
        det_thresh_, max_age_, min_hits_, iou_threshold_, delta_t_,
        "iou", inertia_, use_byte_);

    // Copy input shape from detector
    input_shape_ = detector_->get_input_shape();
    model_loaded_ = true;
}

cv::Mat OCSortTracker::preprocess(const cv::Mat& image) {
    ensure_model_loaded();
    return detector_->preprocess(image);
}

vision_common::TrackingResultList OCSortTracker::track(
    const cv::Mat& image,
    float conf_threshold,
    float iou_threshold) {
    vision_core::ImageInput input;
    input.image = image;
    return track_input(input, conf_threshold, iou_threshold);
}

vision_common::TrackingResultList OCSortTracker::track_input(
    const vision_core::ImageInput& input,
    float conf_threshold,
    float iou_threshold) {
    ensure_model_loaded();
    reset_runtime_profile();
    const auto t0 = std::chrono::steady_clock::now();

    // Run detection
    const auto t_det0 = std::chrono::steady_clock::now();
    vision_common::DetectionResultList detections =
        detector_->detect_input(
            input, conf_threshold, iou_threshold);
    const auto t_det1 = std::chrono::steady_clock::now();
    set_runtime_detect_ms(std::chrono::duration<double, std::milli>(t_det1 - t_det0).count());

    // Convert to Eigen matrix format for OC-SORT tracker
    const auto t_track0 = std::chrono::steady_clock::now();
    Eigen::MatrixXf dets = convert_results_to_dets(detections);

    // Update tracker
    std::vector<Eigen::RowVectorXf> tracks = tracker_->update(dets);

    // Convert back to TrackingResult format with track_id and preserve label information
    vision_common::TrackingResultList results = convert_tracks_to_results(tracks, detections);
    const auto t_track1 = std::chrono::steady_clock::now();
    set_runtime_track_ms(std::chrono::duration<double, std::milli>(t_track1 - t_track0).count());

    const auto t1 = std::chrono::steady_clock::now();
    set_runtime_total_ms(std::chrono::duration<double, std::milli>(t1 - t0).count());

    return results;
}


std::vector<vision_core::InferIntent> OCSortTracker::supported_intents() const {
    return {vision_core::InferIntent::kTrack};
}

vision_core::InferResponse OCSortTracker::Run(const vision_core::InferRequest& request) {
    assert(request.intent == vision_core::InferIntent::kTrack);
    const auto* image_input = std::get_if<vision_core::ImageInput>(&request.input);
    if (image_input == nullptr) {
        vision_core::InferResponse response;
        response.ok = false;
        response.error_message = "OCSortTracker expects ImageInput";
        return response;
    }

    vision_common::TrackingResultList task_results =
        track_input(
            *image_input,
            request.params.conf_threshold,
            request.params.iou_threshold);
    vision_core::InferResponse response;
    response.results.reserve(task_results.size());
    for (auto& item : task_results) {
        response.results.emplace_back(std::move(item));
    }
    return response;
}

std::vector<vision_core::ModelCapability> OCSortTracker::get_capabilities() const {
    return {vision_core::ModelCapability::kDraw};
}

void OCSortTracker::configure_preprocess_backend(
    const std::string& backend)
{
    BaseModel::configure_preprocess_backend(backend);
    detector_->configure_preprocess_backend(backend);
}

Eigen::MatrixXf OCSortTracker::convert_results_to_dets(
    const vision_common::DetectionResultList& results) {
    if (results.empty()) {
        return Eigen::MatrixXf(0, 6);
    }

    // Each row: [x1, y1, x2, y2, score, label]
    Eigen::MatrixXf dets(results.size(), 6);

    for (size_t i = 0; i < results.size(); ++i) {
        dets(i, 0) = results[i].bbox.x1;
        dets(i, 1) = results[i].bbox.y1;
        dets(i, 2) = results[i].bbox.x2;
        dets(i, 3) = results[i].bbox.y2;
        dets(i, 4) = results[i].score;
        dets(i, 5) = static_cast<float>(results[i].label);
    }

    return dets;
}

float OCSortTracker::calculate_iou(float x1_1, float y1_1, float x2_1, float y2_1,
                                    float x1_2, float y1_2, float x2_2,
                                    float y2_2) {
    float inter_x1 = std::max(x1_1, x1_2);
    float inter_y1 = std::max(y1_1, y1_2);
    float inter_x2 = std::min(x2_1, x2_2);
    float inter_y2 = std::min(y2_1, y2_2);

    if (inter_x2 <= inter_x1 || inter_y2 <= inter_y1) {
        return 0.0f;
    }

    float inter_area = (inter_x2 - inter_x1) * (inter_y2 - inter_y1);
    float area1 = (x2_1 - x1_1) * (y2_1 - y1_1);
    float area2 = (x2_2 - x1_2) * (y2_2 - y1_2);
    float union_area = area1 + area2 - inter_area;

    if (union_area <= 0.0f) {
        return 0.0f;
    }

    return inter_area / union_area;
}

vision_common::TrackingResultList OCSortTracker::convert_tracks_to_results(
    const std::vector<Eigen::RowVectorXf>& tracks,
    const vision_common::DetectionResultList& detections) {

    vision_common::TrackingResultList results;
    results.reserve(tracks.size());

    for (const auto& track : tracks) {
        vision_common::TrackingResult result;

        // OC-SORT output format: [x1, y1, x2, y2, track_id, ...]
        // Note: The exact format depends on OC-SORT implementation
        if (track.size() >= 5) {
            result.bbox = vision_common::BoundingBox{
                static_cast<float>(track(0)),
                static_cast<float>(track(1)),
                static_cast<float>(track(2)),
                static_cast<float>(track(3))
            };
            result.track_id = static_cast<int>(track(4));

            // Default score (OC-SORT may not preserve score)
            result.score = 1.0f;
            result.state = vision_common::TrackingResult::State::Confirmed;

            // Match with detections to find label using IoU
            float max_iou = 0.0f;
            int best_match_idx = -1;

            for (size_t i = 0; i < detections.size(); ++i) {
                float iou = vision_common::iou(result.bbox, detections[i].bbox);
                if (iou > max_iou && iou > 0.5f) {
                    max_iou = iou;
                    best_match_idx = static_cast<int>(i);
                }
            }

            // Use label from matched detection, or -1 if no good match
            if (best_match_idx >= 0) {
                result.label = detections[best_match_idx].label;
                result.score = detections[best_match_idx].score;
            } else {
                result.label = -1;
            }

            results.push_back(result);
        }
    }

    return results;
}

// Self-registration (runs at program startup)
static vision_core::ModelRegistrar<OCSortTracker> registrar("OCSortTracker");

}  // namespace vision_deploy
